File: 0933-number-of-recent-calls/0933-number-of-recent-calls.cpp
class RecentCounter {
    queue<int> q;
    int counter;
public:
    RecentCounter() {
        counter = 0;
    }
    
    int ping(int t) {
        while (!q.empty() && q.front() < t - 3000) {
            q.pop();
            counter--;
        }
        q.push(t);
        counter++;
        return counter;
    }
};

/**
 * Your RecentCounter object will be instantiated and called as such:
 * RecentCounter* obj = new RecentCounter();
 * int param_1 = obj->ping(t);
 */

File: 0933-number-of-recent-calls/0933-number-of-recent-calls.java
class RecentCounter {
    ArrayDeque<Integer> q;
    int counter;
    public RecentCounter() {
        this.q = new ArrayDeque<>();
    }
    
    public int ping(int t) {
        int expiredRequsts = 0;
        // Remove all expired pings (pings < t - 3000) from front (Dequeue)
        while (!q.isEmpty() && q.peek() < t - 3000) {
            q.poll();
            counter--;
        }
        q.offer(t);
        counter++;
        return counter;

    }
}

/**
 * Your RecentCounter object will be instantiated and called as such:
 * RecentCounter obj = new RecentCounter();
 * int param_1 = obj.ping(t);
 */